---
title: "SEESAW: Statistical Estimation Of Allelic Expression Using Salmon And Swish"
author: Euphy Wu, Michael Love
date: "July 2022"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{SEESAW: Statistical Estimation Of Allelic Expression Using Salmon And Swish}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Overview

### Description

In this workshop, we will cover a newly developed suite of tools for
performing analysis of allelic expression in Bioconductor. You will
learn about how we use upstream tools like `g2gtools` and `Salmon` to
quantify allelic expression for RNA-seq, and then how to explore the
data in R using the `fishpond` package. Altogether we call this suite
of tools *SEESAW* for
"Statistical Estimation Of allelic Expression Using Salmon And sWish".

### Background and other methods

The SEESAW set of methods for quantifying and analyzing allelic data
is similar to 
[mmseq](https://github.com/eturro/mmseq), 
in which uncertainty of assignment of reads to isoforms and alleles is
estimated, stored, and using in statistical testing.
It is also similar to 
[EMASE](https://emase.readthedocs.io/en/latest/index.html), which
provides quantification at isoform and allele level;
SEESAW also use the same upstream pipeline as EMASE for construction
of diploid reference transcripts.

Our approach differs from existing methods for detecting allelic
imbalance in RNA-seq data such as 
[WASP](https://github.com/bmvdgeijn/WASP)
or the Bioconductor package
[AllelicImbalance](https://www.bioconductor.org/packages/AllelicImbalance),
in that we do not focus on pileup of reads on particular SNPs, but
instead integrate the isoform- and allele-level information contained
in reads aligning to entire transcripts (including both SNP and indel
variation). However, these existing methods can be used across
patients with different genomes, while the SEESAW methods are designed
for analyzing replicate samples with the same diploid genomes (e.g. F1
crosses, or time series of human cell lines).

A recent method for isoform- and allele-level testing is the
Bioconductor package
[PAIRADISE](https://www.bioconductor.org/packages/PAIRADISE),
which quantifies reads that contain both allelic and isoform-level
information: those that contain both a heterozygous SNP and a splice
juntion.
Another relevant recently developed method is
[ASEP](https://github.com/Jiaxin-Fan/ASEP),
which allows for gene-level allelic imbalance detection, accounting
for unknown genotype of other nearby regulatory variants.
PAIRADISE and ASEP are therefore applicable across populations of
genotypically diverse individuals. We link to these publications for
background reading below.

### Pre-requisites

* Some basic background of RNA biology, e.g. what are genes, what are isoforms/transcripts?
* Some familiarity with RNA-seq data, e.g. what are reads?
* Some familiarity with working with *SummarizedExperiment* objects will help.

### Background reading

* [Tools and best practices for data processing in allelic expression analysis](https://doi.org/10.1186/s13059-015-0762-6)
* [A vast resource of allelic expression data spanning human tissues](https://doi.org/10.1186/s13059-020-02122-z)
* [Haplotype and isoform specific expression estimation using multi-mapping RNA-seq reads](https://doi.org/10.1186/gb-2011-12-2-r13) (mmseq)
* [Hierarchical analysis of RNA-seq reads improves the accuracy of allele-specific expression](https://doi.org/10.1093/bioinformatics/bty078) (EMASE)
* [Detecting Allele-Specific Alternative Splicing from Population-Scale RNA-Seq Data](https://doi.org/10.1016/j.ajhg.2020.07.005) (PAIRADISE)
* [ASEP: Gene-based detection of allele-specific expression across individuals in a population by RNA sequencing](https://doi.org/10.1371/journal.pgen.1008786)

### Participation

The format is lecture + lab, with time for data exploration.

### _R_ / _Bioconductor_ packages used

* SummarizedExperiment
* fishpond
* plyranges
* pheatmap
* Gviz

### Time outline

An example for a 45-minute workshop:

| Activity                     | Time |
|------------------------------|------|
| Intro to allelic quant       | 15m  |
| Exploration of data in R     | 15m  |
| Testing and visualization    | 15m  |

### Workshop goals and objectives

### Learning goals

* understand why analyzing allelic expression can be useful for
  learning about *cis* genetic regulation
* understand differences between approaches to allelic quantification
* understand difference among types of allelic imbalance tests,
  e.g. "global", "differential", and "dynamic" imbalance

### Learning objectives

* import allelic quantification data into R
* exploratory analysis of allelic data with uncertainty
* statistical testing for allelic imbalance across samples
* exploration of testing results with heatmaps and genome browser
  plots 

## What is allelic expression

```{r, echo=FALSE, fig.alt="Diagram of chromosomes with homozygous and heterozygous alleles, from URL: https://commons.wikimedia.org/wiki/File:Heterozygous.jpg"}
knitr::include_graphics("images/heterozygous.jpg")
```

Let us assume for this workshop that we are studying a diploid
organism, and we focus on the genes with two copies (so genes on the
autosome, and e.g. chrX for XX individuals). We do not consider here
allelic quantification in cases of complex copy number such as tumor
gene expression. 

In most RNA-seq pipelines, we are concerned with *total expression* of
genes or transcripts, where we do not distinguish between RNA
transcripts from either of the two chromosomes. Suppose the light blue
region highlighted above is a gene, and there are two alleles (B and
b) which differ by one or more genetic variants. In the case that
there are genetic variants falling within the exons of the gene, some
RNA-seq reads may contain information that allows us to determine the
origin of the RNA transcript, whether from the B chromosome or the b
chromosome. Typically, RNA-seq analysis pipelines just are concerned
with counting reads aligning to the light blue region, regardless of
which chromosome. Some pipelines will attempt to identify which splice
isoforms these align to, but do not attempt to distinguish between
alleles. 

In SEESAW (and mmseq and EMASE) we attempt to distinguish both allele
and isoform from the RNA-seq reads. Again, note that, while there is
always expression from two alleles, only in the middle two cases
above, with one or genetic variants falling in the exons of the
highlighted gene, can we recover allelic information from RNA-seq
reads. In this workshop, we will therefore look at two counts per
sample, one count for the `a2` or reference allele, and the other
count for the `a1` or alternate allele. In our case, the reference
will be the C57BL/6J (or "B6") mouse genome, and the alternate allele
will be the CAST/EiJ mouse genome (we will focus on the B6xCAST F1
hybrid mouse). The mouse pre-osteoblast like cells went through
differentiation induced with an osteoblast differentiation cocktail
and RNA was collected every two days from day 2 to day 18 post
differentiation (nine time points).

## Allelic imbalance

The expression of the two alleles may differ for a variety of
reasons. Imagine a single cell with two copies of a gene that may be
transcribed. You may be familiar with the concept of "transcriptional
bursting":

> Mammalian gene expression is inherently stochastic, and results
> in discrete bursts of RNA molecules that are synthesized from each
> allele.

-Larsson *et al.* "Genomic encoding of transcriptional burst
kinetics." *Nature* **565**, 251â€“254 (2019) 
<https://doi.org/10.1038/s41586-018-0836-1>.
(Other papers describing the stochastic nature of expression from each
allele are the 
[SCALE](https://doi.org/10.1186/s13059-017-1200-8) and 
[scBASE](https://doi.org/10.1038/s41467-019-13099-0) 
method papers.)

We might expect that these stochastic differences between the two
alleles would average out if we look at many cells, as in bulk
RNA-seq. But in some cases, we may see systematic differences between
the two alleles. Some reasons why:

1. Interesting biological differences
2. Technical artifacts

As in 
[Figure
1](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0762-6/figures/1)
of Castel *et al.* (2015), the largest differences seen in allelic
imbalance are likely mapping bias or genotyping errors (e.g. #2
above). If we map reads to the reference genome, then we will
generate a systematic bias where reads from the alternate allele will
tend to have lower counts. Likewise, if one of the two genomes has
lower quality, even if we map reads to both genomes or transcriptomes
(as in SEESAW) we will see a systematic bias towards the higher
quality genome. Here we assume both genomes are equally free of
errors, but assessing systematic bias is an important part of allelic
analysis (e.g. histograms or MA plots of the difference between the
mapping rate of the two alleles across all genes).

However, of the type #1 above, there are some possibilities for the
source of allelic differences in RNA abundance:

* Epigenetic (e.g. the common example of chrX inactivation)
* Nonsense-mediated decay (NMD)
* *Cis*-genetic regulation (cGR)

In the first case, one allele is genomically imprinted, which may
involve DNA methylation and histone modifications depending on the
species. In the second and third case, RNA is often produced from both
alleles, but one is degraded at a higher rate (NMD) or produced at a
higher rate (cGR). 

Our motivation in developing SEESAW was particular on detecting cGR,
as we are interested in how non-coding variants *cis* to a gene may
affect transcription. These could affect transcription through
alteration of transcription factor binding sites (TFBS) or through
changes to splicing. While both TFBS and splicing changes are of
interest, in this workshop we will focus on non-coding variation that
affects TFBS and therefore modulates transcriptional activity at the
promoter.

We therefore will focus on cases where there is allelic imbalance of a
transcript, and we can detect this imbalance with RNA-seq (center two
cases in the above diagram), and we hypothesize that this imbalance is
the result of a heterozygous variant in some regulatory region of the
gene. As we only focus here on RNA-seq, we will not have additional
data that allows us to locate the source of the regulatory change
across the two alleles. For this, it would be useful to have
complementary ATAC-seq or ChIP-seq (if a particular TF is of
interest).

## SEESAW pipeline

At the beginning of the workshop, we will give an overview of the
steps that need to take place outside of R, before we import the
allelic data into Bioconductor. These are outlined as the first parts
of the pipeline in the diagram below. These are also described in
detail in the 
[allelic vignette](https://mikelove.github.io/fishpond/articles/allelic.html) 
for the fishpond package.

The input to the pipeline of methods is RNA-seq data (FASTQ files),
information about the diploid reference genome (e.g. VCF file), and
information about the isoforms (e.g. GTF file). The SEESAW suite of
methods is currently designed to work for experiments with multiple
replicates of samples with the same diploid genome, e.g. F1 crosses,
or time series of human cell lines.

The R code below will begin at the center blue square with
`importAllelicCounts()`, assuming that diploid quantification with
bootstrap replicates for uncertainty has already been performed.

```{r, echo=FALSE, fig.alt="Diagram of SEESAW pipeline, including g2gtools, Salmon index, Salmon quant, and import into Bioconductor using fishpond package funcctions"}
knitr::include_graphics("images/SEESAW.png")
```

Links to relevant software: 

* [g2gtools](http://churchill-lab.github.io/g2gtools/) for
  constructing the diploid transcriptome
* Salmon, run with [bootstrap inferential replicates](https://salmon.readthedocs.io/en/latest/salmon.html#numbootstraps)

## Importing allelic counts

The following five un-evaluated code chunks were used to import the
allelic quantification from Salmon into R. These are un-evaluated, as
the original data is too large to include in this package.

First, we define `coldata` and `txps`, which describe the sample
metadata, and the genomic ranges of the transcripts that were
quantified (the haploid reference transcripts). We do not yet have a
convenient way to describe the location of the transcripts in the
diploid genome, so we still rely on a single reference genome here,
although a diploid genome was used for quantifying allelic expression.

```{r eval=FALSE}
library(AnnotationHub)
library(ensembldb)
# `files` points to `quant.sf` files in original Salmon directories
# other vectors here describe the samples
coldata <- data.frame(cross, day, files, names)
ah <- AnnotationHub()
#query(ah, c("EnsDb","102","Mus musculus"))
edb <- ah[["AH89211"]] # download annotation for mouse from Ensembl 
txps <- transcripts(edb)
```

This is how a sample `coldata` looks like:

|   |  cross  | day |   files             |    names    |
|---|---------|-----|---------------------|-------------|
| 1 | 129xB6  |  2  | `/path/to/quant.sf` | 129xB6-d02  |
| 2 | 129xB6  |  4  | `/path/to/quant.sf` | 129xB6-d04  |
| 3 | CASTxB6 |  2  | `/path/to/quant.sf` | CASTxB6-d02 |
| 4 | CASTxB6 |  4  | `/path/to/quant.sf` | CASTxB6-d04 |

We can group the transcripts to the gene level with the following `group_id`:

```{r eval=FALSE}
library(plyranges)
txps <- txps %>%
  select(tx_id, group_id=gene_id)
```

The following code chunk imports the allelic data, creating a "wide"
*SummarizedExperiment* object which will have 2x the number of columns
as the samples listed in `coldata`. We specify the strings that
distinguish the `a1` and `a2` allele in the transcript FASTA file 
(these are suffices that follow the transcript name, separated with an
underscore, e.g. `ENST123_ref` and `ENST123_alt`.
By convention, the `a2` allele describes the reference allele.
In the `gse` object (gene-level SummarizedExperiment), the `a2`
allelic counts will appear as the first columns followed by the `a1`
allelic counts. Here we supply `txps` as a *GRanges* object with a
metadata column `group_id` that describes how we collapse
transcript-level counts.


```{r eval=FALSE}
library(fishpond)
gse <- importAllelicCounts(
  coldata, a1="alt", a2="ref",
  format="wide", tx2gene=txps,
)
```

Finally, we performed minimal filtering on the features:

```{r eval=FALSE}
# filtering out lowly expressed features:
keep <- rowSums(assay(gse) >= 10) >= 6
gse <- gse[keep,]
```

Alternatively, we can group transcript-level counts to the
transcription start site (TSS) level
using the following code chunk to define `txps`. `makeTx2Tss()` is a
convenience function in *fishpond*. The `maxgap=50` argument means we
will group together transcripts that have TSS that fall within 50bp of
each other.

```{r eval=FALSE}
txps <- makeTx2Tss(edb, maxgap=50) %>%
  select(tx_id, gene_id, group_id, tss)
```

**Q: why would we want to collapse to TSS level?**

## Mouse osteoblast differentiation

We have two pre-packaged datasets as part of this workshop
package, one where allelic counts from the osteoblast RNA-seq experiment are
summarized to the gene level, and another where they are summarized to
the TSS level. 

The datasets are described in the man pages (see References tab).
Briefly, mouse osteoblast cells were differentiated over a time course
from day 2 to day 18. The data has been subset to just the genes on
chr1. A reference for the experiment is:

> Kemp JP, Medina-Gomez C, Estrada K, St Pourcain B et al.
> Phenotypic dissection of bone mineral density reveals skeletal site specificity
> and facilitates the identification of novel loci in the genetic regulation of
> bone mass attainment. PLoS Genet 2014 Jun;10(6):e1004423. PMID: 24945404
> <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4063697/>

The RNA-seq reads were processed via the quantification part of
*SEESAW* as described above. Only data from chromosome 1 is included
in the pre-packaged datasets for simplicity of the workshop.

**Q: why would we be interested in allelic expression during
osteoblast differentiation?**

## Explore allelic counts

We first load the gene-level dataset created with `importAllelicCounts()`:

```{r message=FALSE}
library(SummarizedExperiment)
library(Bioc2022AllelicExpression)
data(osteoblast_gene_chr1)
osteoblast_gene_chr1
g <- osteoblast_gene_chr1
```

The sample metadata:

```{r}
colData(g)
```

For each sample and each allele, we have counts and also bootstrap
replicates. The bootstrap replicates are from Salmon, by specifying
`--numBootstraps 30` during quantification. 

**Q: Why do we need bootstrap replicates?**

We can use *fishpond* functions to do some basic exploration of the
allelic counts. 

The *counts* assay provides the read counts for features x samples, as
Salmon is used for quantification these are estimated counts that are
non-negative but fractional, not integer. The *abundance* assay gives
estimates in Transcripts Per Million (TPM). The *length* assay gives
the average effective transcript length for a given transcript or a
given gene. Finally the *infRep*'s are the bootstrap samples for the
counts.

```{r}
library(fishpond)
assayNames(g)
```

`getTrace` obtains a trace of the inferential replicates of counts for
one feature and one sample (multiple features/samples can also be
returned). Plotting the trace in the histogram helps us to visualize
the distribution of the estimated counts for the day 2 sample of the
CASTxB6 cross, reference allele (`a2`). We can see for this particular
sample, the counts have roughly the same mean and variance, indicating
that there is not much uncertainty (minimal bootstrap variance will
roughly follow what is expected for a Poisson random variable).

```{r histBoot}
dat <- getTrace(g, idx=1, samp_idx="CASTxB6-d02-a2")
hist(dat$count, border="white", col="grey50", xlab="estimated count, a2")
c(mu=mean(dat$count), var=var(dat$count))
```

We can also visualize the bootstrap distributions for all samples for
a given gene using `plotInfReps()`. Here we have a boxplot of
inferential replicates for gene `ENSMUSG00000004110`, grouped by
allele.

```{r plotInfReps}
plotInfReps(g, idx=1, x="allele", legend=TRUE)
```

We can also group the allelic counts ordered by sample, e.g.:

```{r plotInfReps2}
gcast <- g[,g$cross == "CASTxB6"]
plotInfReps(gcast, idx=1, x="allele", cov="day", legend=TRUE)
```

In the above plot, the samples are ordered by sample, as if `day` x
`allele` were a categorical variable. We can also plot `day` on its
original continuous scale. To do so, we use points and lines to
represent the bootstrap distribution instead of boxplots. To draw the
line, we need to first compute the bootstrap variance, which can be
computed with the `computeInfRV()` function (InfRV = "inferential
relative variance", a useful diagnostic for comparing uncertainty
across features):

```{r plotInfReps3}
gcast <- computeInfRV(gcast)
plotInfReps(gcast, idx=1, x="day", cov="allele",
            shiftX=.2, legend=TRUE)
```

## Filter non-informative features

Some features have no sequence difference between the two alleles. One
easy way to detect these is to look at a single bootstrap replicate,
e.g. the first one `infRep1`. If this bootstrap sample gives the same
count to each allele, across all samples, this implies either that the
sequence was identical across alleles, or there were no reads covering
the part of the transcript sequence that contains a difference across
alleles. We remove these from the dataset, as they are non-informative
for allelic imbalance analysis.

```{r}
n <- ncol(gcast)/2
rep1a1 <- assay(gcast, "infRep1")[, gcast$allele == "a1" ]
rep1a2 <- assay(gcast, "infRep1")[, gcast$allele == "a2" ]
someInfo <- rowSums(abs(rep1a1 - rep1a2) < 1) < n
table(someInfo)
gcast <- gcast[ someInfo, ]
```

## Allelic imbalance by TSS Group 

### Running Swish at TSS level

Here we switch to using TSS-level data. We have grouped transcripts
together by TSS using +/- 50bp tolerance (so forming fewer groups by
allowing +/- 50bp between TSS to join them into a group).

We first filter out non-informative features as shown in the previous
step: 

```{r}
data(osteoblast_tss_chr1)
tss <- osteoblast_tss_chr1
tcast <- tss[,tss$cross == "CASTxB6"]
n <- ncol(tcast)/2
rep1a1 <- assay(tcast, "infRep1")[, tcast$allele == "a1" ]
rep1a2 <- assay(tcast, "infRep1")[, tcast$allele == "a2" ]
someInfo <- rowSums(abs(rep1a1 - rep1a2) < 1) < n
table(someInfo)
tcast <- tcast[ someInfo, ]
```

Running `swish` generally involves scaling the inferential replicates,
filtering out rows with insufficient counts, and calculating the
statistics (log fold changes, test statistics, p-values and
q-values). 
Here, we skip scaling the counts (to account for sequencing depth)
since we are comparing the counts of the two alleles within
samples. By default, `labelKeep` will label features with a minimal
count of 10 in at least 3 samples.

Here we are performing two types of allelic imbalance testing.
*Global AI* tests for consistent allelic imbalance across all
samples. 
*Dynamic AI* tests for non-zero correlation between the log allelic
fold change and a continuous variable (in this case, the `day`).

```{r}
tcast <- labelKeep(tcast)
tcast <- tcast[mcols(tcast)$keep,]
# for plots later
tcast <- computeInfRV(tcast)
```

The tests involve permutation, and so it takes about 20 seconds in
this case.

```{r}
set.seed(1)
global <- swish(tcast, x="allele", pair="day")
```

The dynamic test here takes about 40 seconds.

```{r}
set.seed(1)
dy <- swish(tcast, x="allele", pair="day",
            cov="day", cor="pearson")
```

**Q: why do we specify `day` twice in the dynamic analysis?**

The results are stored in `mcols()` in each case. We can examine how
many TSS groups are detected as having global or dynamic AI in a 5%
FDR set:

```{r}
table(mcols(global)$qvalue < 0.05)
table(mcols(dy)$qvalue < 0.05)
```

**Q: why are there fewer dynamic positive results?**

### Plotting Results

We can visualize the relationship between allelic log fold change
versus mean expression through MA plots. Blue points are TSS groups in
the 5% FDR set. One can specify the threshold by using `alpha = ...`
option.

```{r maGlobal}
plotMASwish(global, alpha=.05, main="Global MA Plot")
```

The average LFC over all samples is plotted also for the dynamic test,
although the correlation is the test statistic used for computing
p-values and q-values. Hence the most extreme LFC are not in the 5%
FDR set.

```{r maDynamic}
plotMASwish(dy, alpha=.05, main="Dynamic MA Plot")
```

#### Global AI visualization

We now want to visualize the TSS group analysis in a genomic
context. To do so, first we must download the mouse Ensembl
annotations. We do so using *AnnotationHub*:

```{r message=FALSE}
library(AnnotationHub)
library(ensembldb)
```

```{r}
ah <- AnnotationHub()
edb <- ah[["AH89211"]]
```

The `plotAllelicGene()` function in `fishpond` can be used to visualize
Swish test results ($-log_{10}(q-value)$ and log2FC), allelic proportions
and isoform proportions (calculated from TPM) within the context of
the gene model. One can also indicate which gene to plot using gene
symbol, e.g. `symbol="Soat1"`. Either *TxDb* or *EnsDb* can be use to
include the gene model at the top of the plot.

```{r globalGene}
plotAllelicGene(global, gene="ENSMUSG00000026600",
                db=edb, labels=list(a2="B6",a1="CAST"))
```

**Q: which isoform has higher abundance?**

We can verify the `plotAllelicGene()` figure with `plotInfReps()`. We
can see B6 has a higher count than CAST in majority of the
samples. One can customize the label of alleles in the function with
the following code, switching from `a2, a1` to `B6, CAST`.

```{r globalInfReps}
tcast$allele_new <- tcast$allele
levels(tcast$allele_new) <- c("B6", "CAST")
idx <- match("ENSMUSG00000026600", mcols(tcast)$gene_id)
plotInfReps(tcast, idx=idx, x="allele_new", cov="day",
            legend=TRUE, legendPos="topright")
```

We can also use a heatmap to visualize the difference in expression
across alleles for the TSS groups, using `plotAllelicHeatmap()`, which
itself calls `pheatmap()`. Darker green in the annotation bar
represents a more significant *qvalue* in differential testing
(for allelic imbalance). A color toward red indicates more of the `a1`
allele (CAST) while a color toward blue indicates more of `a2` (B6).

```{r globalHeatmap}
idx <- mcols(global)$gene_id == "ENSMUSG00000026600"
dat <- data.frame(minusLogQ=-log10(mcols(global)$qvalue[idx]),
                  row.names=rownames(global)[idx])
colnames(global) <- c(paste0("Day ",seq(2,18, by= 2), "-a2"), 
                  paste0("Day ",seq(2,18, by= 2), "-a1"))
plotAllelicHeatmap(global, idx=idx, annotation_row=dat)
```

#### Dynamic AI visualization

`plotInfReps()` is a useful function for visualizing how allelic
counts changes over a continuous covariate, such as `day`. Using
`lines` can help us better visualize the trend.

```{r dynamicInfReps1}
plotCurves <- function(dy, idx) {
  cols <- c("dodgerblue","goldenrod4")
  plotInfReps(dy, idx=idx, x="day", cov="allele", shiftX=0.01)
  lines(dy$day[1:9], assay(dy, "mean")[idx,1:9], col=cols[1]) 
  lines(dy$day[10:18], assay(dy, "mean")[idx,10:18], col=cols[2])
  legend("topleft", legend=c("B6", "CAST"), pch=15, col=cols, inset=.05)
}
idx <- "ENSMUSG00000026185-1"
plotCurves(dy, idx)
```

```{r dynamicInfReps2}
idx <- "ENSMUSG00000026185-2"
plotCurves(dy, idx)

```

**Q: What do you notice with the two infRV plots? HINT: Is the allelic
imbalance in the same direction?**

`plotAllelicGene()` can also be used to visualize time-series allelic
imbalance data. We first need to group the timepoints (or else the
plot would have no room as there are 9 distinct time points). In this
example, we are grouping Day 2-6, Day 8-12, and Day 14-18
together. How best to group the timepoints depends on either the
biological system of a strategy to demonstrate maximal differences in
the ratio per group (e.g. by looking at `plotInfReps()` plot). 

The plot shows allelic proportions with grouped timepoints, as well as
isoform proportion. We can see an increasing in B6 expression over
time for the first transcript group and an increasing in CAST
expression with time for the second transcript group.

```{r dynamicGene}
# this makes integer bins
dy$time_bins <- cut(dy$day,breaks=c(2,6,12,18),
                    include.lowest=TRUE, labels=FALSE)
# come up with useful time group labels
time_labels <- c("Day02-06","Day08-12","Day14-18")
# use new time group labels
dy$time_bins <- time_labels[ dy$time_bins ]
gene <- "ENSMUSG00000026185"
plotAllelicGene(dy, gene=gene, db=edb, cov="time_bins",
                qvalue=FALSE, log2FC=FALSE,
                tpmFilter=.1, isoPropFilter=.01,
                labels=list(a2="B6",a1="CAST"))
```

The same gene, but looking at `infReps`:

```{r dynamicInfReps3}
idx <- which(mcols(dy)$gene_id == gene)
par(mfrow=c(2,1), mar=c(3,4.5,1.5,1))
plotCurves(dy, idx[1])
plotCurves(dy, idx[2])
```

**Q: What type of plot might help to visualize time-series allelic data
across multiple transcripts and multiple samples?**

We can add a time bar on the top of a heatmap to indicate time by
defining `col_dat`.

```{r dynamicHeatmap}
idx <- match("ENSMUSG00000026185", mcols(dy)$gene_id)
dat <- data.frame(minusLogQ=-log10(mcols(dy)$pvalue[idx]),
                  row.names=rownames(dy)[idx])
colnames(dy) <- c(paste0("Day ",seq(2,18, by= 2), "-a2"), 
                  paste0("Day ",seq(2,18, by= 2), "-a1"))
row_dat <- data.frame(minusLogQ=-log10(mcols(dy)$qvalue[idx]),
                      row.names=rownames(dy)[idx])
col_dat <- data.frame(time=dy$day[1:9],
                      row.names=paste0("Day ",seq(2,18, by= 2)))
plotAllelicHeatmap(dy, idx=idx,
                   annotation_row=row_dat,
                   annotation_col=col_dat,
                   cluster_rows=FALSE)

```
