<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>SEESAW: Statistical Estimation Of Allelic Expression Using Salmon And Swish • Bioc2022AllelicExpression</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="SEESAW: Statistical Estimation Of Allelic Expression Using Salmon And Swish">
<meta property="og:description" content="Bioc2022AllelicExpression">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Bioc2022AllelicExpression</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/Bioc2022AllelicExpression.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mikelove/Bioc2022AllelicExpression/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>SEESAW: Statistical Estimation Of Allelic
Expression Using Salmon And Swish</h1>
                        <h4 data-toc-skip class="author">Euphy Wu,
Michael Love</h4>
            
            <h4 data-toc-skip class="date">July 2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/mikelove/Bioc2022AllelicExpression/blob/HEAD/vignettes/Bioc2022AllelicExpression.Rmd" class="external-link"><code>vignettes/Bioc2022AllelicExpression.Rmd</code></a></small>
      <div class="hidden name"><code>Bioc2022AllelicExpression.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<div class="section level3">
<h3 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h3>
<p>In this workshop, we will cover a newly developed suite of tools for
performing analysis of allelic expression in Bioconductor. You will
learn about how we use upstream tools like <code>g2gtools</code> and
<code>Salmon</code> to quantify allelic expression for RNA-seq, and then
how to explore the data in R using the <code>fishpond</code> package.
Altogether we call this suite of tools <em>SEESAW</em> for “Statistical
Estimation Of allelic Expression Using Salmon And sWish”.</p>
</div>
<div class="section level3">
<h3 id="background-and-other-methods">Background and other methods<a class="anchor" aria-label="anchor" href="#background-and-other-methods"></a>
</h3>
<p>The SEESAW set of methods for quantifying and analyzing allelic data
is similar to <a href="https://github.com/eturro/mmseq" class="external-link">mmseq</a>, in
which uncertainty of assignment of reads to isoforms and alleles is
estimated, stored, and using in statistical testing. It is also similar
to <a href="https://emase.readthedocs.io/en/latest/index.html" class="external-link">EMASE</a>,
which provides quantification at isoform and allele level; SEESAW also
use the same upstream pipeline as EMASE for construction of diploid
reference transcripts.</p>
<p>Our approach differs from existing methods for detecting allelic
imbalance in RNA-seq data such as <a href="https://github.com/bmvdgeijn/WASP" class="external-link">WASP</a> or the Bioconductor
package <a href="https://www.bioconductor.org/packages/AllelicImbalance" class="external-link">AllelicImbalance</a>,
in that we do not focus on pileup of reads on particular SNPs, but
instead integrate the isoform- and allele-level information contained in
reads aligning to entire transcripts (including both SNP and indel
variation). However, these existing methods can be used across patients
with different genomes, while the SEESAW methods are designed for
analyzing replicate samples with the same diploid genomes (e.g. F1
crosses, or time series of human cell lines).</p>
<p>A recent method for isoform- and allele-level testing is the
Bioconductor package <a href="https://www.bioconductor.org/packages/PAIRADISE" class="external-link">PAIRADISE</a>,
which quantifies reads that contain both allelic and isoform-level
information: those that contain both a heterozygous SNP and a splice
juntion. Another relevant recently developed method is <a href="https://github.com/Jiaxin-Fan/ASEP" class="external-link">ASEP</a>, which allows for
gene-level allelic imbalance detection, accounting for unknown genotype
of other nearby regulatory variants. PAIRADISE and ASEP are therefore
applicable across populations of genotypically diverse individuals. We
link to these publications for background reading below.</p>
</div>
<div class="section level3">
<h3 id="pre-requisites">Pre-requisites<a class="anchor" aria-label="anchor" href="#pre-requisites"></a>
</h3>
<ul>
<li>Some basic background of RNA biology, e.g. what are genes, what are
isoforms/transcripts?</li>
<li>Some familiarity with RNA-seq data, e.g. what are reads?</li>
<li>Some familiarity with working with <em>SummarizedExperiment</em>
objects will help.</li>
</ul>
</div>
<div class="section level3">
<h3 id="background-reading">Background reading<a class="anchor" aria-label="anchor" href="#background-reading"></a>
</h3>
<ul>
<li><a href="https://doi.org/10.1186/s13059-015-0762-6" class="external-link">Tools and best
practices for data processing in allelic expression analysis</a></li>
<li><a href="https://doi.org/10.1186/s13059-020-02122-z" class="external-link">A vast resource
of allelic expression data spanning human tissues</a></li>
<li>
<a href="https://doi.org/10.1186/gb-2011-12-2-r13" class="external-link">Haplotype and
isoform specific expression estimation using multi-mapping RNA-seq
reads</a> (mmseq)</li>
<li>
<a href="https://doi.org/10.1093/bioinformatics/bty078" class="external-link">Hierarchical
analysis of RNA-seq reads improves the accuracy of allele-specific
expression</a> (EMASE)</li>
<li>
<a href="https://doi.org/10.1016/j.ajhg.2020.07.005" class="external-link">Detecting
Allele-Specific Alternative Splicing from Population-Scale RNA-Seq
Data</a> (PAIRADISE)</li>
<li><a href="https://doi.org/10.1371/journal.pgen.1008786" class="external-link">ASEP:
Gene-based detection of allele-specific expression across individuals in
a population by RNA sequencing</a></li>
</ul>
</div>
<div class="section level3">
<h3 id="participation">Participation<a class="anchor" aria-label="anchor" href="#participation"></a>
</h3>
<p>The format is lecture + lab, with time for data exploration.</p>
</div>
<div class="section level3">
<h3 id="r-bioconductor-packages-used">
<em>R</em> / <em>Bioconductor</em> packages used<a class="anchor" aria-label="anchor" href="#r-bioconductor-packages-used"></a>
</h3>
<ul>
<li>SummarizedExperiment</li>
<li>fishpond</li>
<li>plyranges</li>
<li>pheatmap</li>
<li>Gviz</li>
</ul>
</div>
<div class="section level3">
<h3 id="time-outline">Time outline<a class="anchor" aria-label="anchor" href="#time-outline"></a>
</h3>
<p>An example for a 45-minute workshop:</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Intro to allelic quant</td>
<td>15m</td>
</tr>
<tr class="even">
<td>Exploration of data in R</td>
<td>15m</td>
</tr>
<tr class="odd">
<td>Testing and visualization</td>
<td>15m</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="workshop-goals-and-objectives">Workshop goals and objectives<a class="anchor" aria-label="anchor" href="#workshop-goals-and-objectives"></a>
</h3>
</div>
<div class="section level3">
<h3 id="learning-goals">Learning goals<a class="anchor" aria-label="anchor" href="#learning-goals"></a>
</h3>
<ul>
<li>understand why analyzing allelic expression can be useful for
learning about <em>cis</em> genetic regulation</li>
<li>understand differences between approaches to allelic
quantification</li>
<li>understand difference among types of allelic imbalance tests,
e.g. “global”, “differential”, and “dynamic” imbalance</li>
</ul>
</div>
<div class="section level3">
<h3 id="learning-objectives">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h3>
<ul>
<li>import allelic quantification data into R</li>
<li>exploratory analysis of allelic data with uncertainty</li>
<li>statistical testing for allelic imbalance across samples</li>
<li>exploration of testing results with heatmaps and genome browser
plots</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="what-is-allelic-expression">What is allelic expression<a class="anchor" aria-label="anchor" href="#what-is-allelic-expression"></a>
</h2>
<p><img src="images/heterozygous.jpg" title="Diagram of chromosomes with homozygous and heterozygous alleles, from URL: https://commons.wikimedia.org/wiki/File:Heterozygous.jpg" alt="Diagram of chromosomes with homozygous and heterozygous alleles, from URL: https://commons.wikimedia.org/wiki/File:Heterozygous.jpg"></p>
<p>Let us assume for this workshop that we are studying a diploid
organism, and we focus on the genes with two copies (so genes on the
autosome, and e.g. chrX for XX individuals). We do not consider here
allelic quantification in cases of complex copy number such as tumor
gene expression.</p>
<p>In most RNA-seq pipelines, we are concerned with <em>total
expression</em> of genes or transcripts, where we do not distinguish
between RNA transcripts from either of the two chromosomes. Suppose the
light blue region highlighted above is a gene, and there are two alleles
(B and b) which differ by one or more genetic variants. In the case that
there are genetic variants falling within the exons of the gene, some
RNA-seq reads may contain information that allows us to determine the
origin of the RNA transcript, whether from the B chromosome or the b
chromosome. Typically, RNA-seq analysis pipelines just are concerned
with counting reads aligning to the light blue region, regardless of
which chromosome. Some pipelines will attempt to identify which splice
isoforms these align to, but do not attempt to distinguish between
alleles.</p>
<p>In SEESAW (and mmseq and EMASE) we attempt to distinguish both allele
and isoform from the RNA-seq reads. Again, note that, while there is
always expression from two alleles, only in the middle two cases above,
with one or genetic variants falling in the exons of the highlighted
gene, can we recover allelic information from RNA-seq reads. In this
workshop, we will therefore look at two counts per sample, one count for
the <code>a2</code> or reference allele, and the other count for the
<code>a1</code> or alternate allele. In our case, the reference will be
the C57BL/6J (or “B6”) mouse genome, and the alternate allele will be
the CAST/EiJ mouse genome (we will focus on the B6xCAST F1 hybrid
mouse).</p>
</div>
<div class="section level2">
<h2 id="allelic-imbalance">Allelic imbalance<a class="anchor" aria-label="anchor" href="#allelic-imbalance"></a>
</h2>
<p>The expression of the two alleles may differ for a variety of
reasons. Imagine a single cell with two copies of a gene that may be
transcribed. You may be familiar with the concept of “transcriptional
bursting”:</p>
<blockquote>
<p>Mammalian gene expression is inherently stochastic, and results in
discrete bursts of RNA molecules that are synthesized from each
allele.</p>
</blockquote>
<p>-Larsson <em>et al.</em> “Genomic encoding of transcriptional burst
kinetics.” <em>Nature</em> <strong>565</strong>, 251–254 (2019) <a href="https://doi.org/10.1038/s41586-018-0836-1" class="external-link uri">https://doi.org/10.1038/s41586-018-0836-1</a>. (Other papers
describing the stochastic nature of expression from each allele are the
<a href="https://doi.org/10.1186/s13059-017-1200-8" class="external-link">SCALE</a> and <a href="https://doi.org/10.1038/s41467-019-13099-0" class="external-link">scBASE</a> method
papers.)</p>
<p>We might expect that these stochastic differences between the two
alleles would average out if we look at many cells, as in bulk RNA-seq.
But in some cases, we may see systematic differences between the two
alleles. Some reasons why:</p>
<ol style="list-style-type: decimal">
<li>Interesting biological differences</li>
<li>Technical artifacts</li>
</ol>
<p>As in <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0762-6/figures/1" class="external-link">Figure
1</a> of Castel <em>et al.</em> (2015), the largest differences seen in
allelic imbalance are likely mapping bias or genotyping errors (e.g. #2
above). If we map reads to the reference genome, then we will generate a
systematic bias where reads from the alternate allele will tend to have
lower counts. Likewise, if one of the two genomes has lower quality,
even if we map reads to both genomes or transcriptomes (as in SEESAW) we
will see a systematic bias towards the higher quality genome. Here we
assume both genomes are equally free of errors, but assessing systematic
bias is an important part of allelic analysis (e.g. histograms or MA
plots of the difference between the mapping rate of the two alleles
across all genes).</p>
<p>However, of the type #1 above, there are some possibilities for the
source of allelic differences in RNA abundance:</p>
<ul>
<li>Epigenetic (e.g. the common example of chrX inactivation)</li>
<li>Nonsense-mediated decay (NMD)</li>
<li>
<em>Cis</em>-genetic regulation (cGR)</li>
</ul>
<p>In the first case, one allele is genomically imprinted, which may
involve DNA methylation and histone modifications depending on the
species. In the second and third case, RNA is often produced from both
alleles, but one is degraded at a higher rate (NMD) or produced at a
higher rate (cGR).</p>
<p>Our motivation in developing SEESAW was particular on detecting cGR,
as we are interested in how non-coding variants <em>cis</em> to a gene
may affect transcription. These could affect transcription through
alteration of transcription factor binding sites (TFBS) or through
changes to splicing. While both TFBS and splicing changes are of
interest, in this workshop we will focus on non-coding variation that
affects TFBS and therefore modulates transcriptional activity at the
promoter.</p>
<p>We therefore will focus on cases where there is allelic imbalance of
a transcript, and we can detect this imbalance with RNA-seq (center two
cases in the above diagram), and we hypothesize that this imbalance is
the result of a heterozygous variant in some regulatory region of the
gene. As we only focus here on RNA-seq, we will not have additional data
that allows us to locate the source of the regulatory change across the
two alleles. For this, it would be useful to have complementary ATAC-seq
or ChIP-seq (if a particular TF is of interest).</p>
</div>
<div class="section level2">
<h2 id="seesaw-pipeline">SEESAW pipeline<a class="anchor" aria-label="anchor" href="#seesaw-pipeline"></a>
</h2>
<p>At the beginning of the workshop, we will give an overview of the
steps that need to take place outside of R, before we import the allelic
data into Bioconductor. These are outlined as the first parts of the
pipeline in the diagram below. These are also described in detail in the
<a href="https://mikelove.github.io/fishpond/articles/allelic.html" class="external-link">allelic
vignette</a> for the fishpond package.</p>
<p>The input to the pipeline of methods is RNA-seq data (FASTQ files),
information about the diploid reference genome (e.g. VCF file), and
information about the isoforms (e.g. GTF file). The SEESAW suite of
methods is currently designed to work for experiments with multiple
replicates of samples with the same diploid genome, e.g. F1 crosses, or
time series of human cell lines.</p>
<p>The R code below will begin at the center blue square with
<code><a href="https://rdrr.io/pkg/fishpond/man/importAllelicCounts.html" class="external-link">importAllelicCounts()</a></code>, assuming that diploid quantification
with bootstrap replicates for uncertainty has already been
performed.</p>
<p><img src="images/SEESAW.png" title="Diagram of SEESAW pipeline, including g2gtools, Salmon index, Salmon quant, and import into Bioconductor using fishpond package funcctions" alt="Diagram of SEESAW pipeline, including g2gtools, Salmon index, Salmon quant, and import into Bioconductor using fishpond package funcctions"></p>
<p>Links to relevant software:</p>
<ul>
<li>
<a href="http://churchill-lab.github.io/g2gtools/" class="external-link">g2gtools</a> for
constructing the diploid transcriptome</li>
<li>Salmon, run with <a href="https://salmon.readthedocs.io/en/latest/salmon.html#numbootstraps" class="external-link">bootstrap
inferential replicates</a>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="importing-allelic-counts">Importing allelic counts<a class="anchor" aria-label="anchor" href="#importing-allelic-counts"></a>
</h2>
<p>The following five un-evaluated code chunks were used to import the
allelic quantification from Salmon into R. These are un-evaluated, as
the original data is too large to include in this package.</p>
<p>First, we define <code>coldata</code> and <code>txps</code>, which
describe the sample metadata, and the genomic ranges of the transcripts
that were quantified (the haploid reference transcripts). We do not yet
have a convenient way to describe the location of the transcripts in the
diploid genome, so we still rely on a single reference genome here,
although a diploid genome was used for quantifying allelic
expression.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># `files` points to `quant.sf` files in original Salmon directories</span></span>
<span><span class="co"># other vectors here describe the samples</span></span>
<span><span class="va">coldata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">cross</span>, <span class="va">day</span>, <span class="va">files</span>, <span class="va">names</span><span class="op">)</span></span>
<span><span class="va">ah</span> <span class="op">&lt;-</span> <span class="fu">AnnotationHub</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#query(ah, c("EnsDb","102","Mus musculus"))</span></span>
<span><span class="va">edb</span> <span class="op">&lt;-</span> <span class="va">ah</span><span class="op">[[</span><span class="st">"AH89211"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">txps</span> <span class="op">&lt;-</span> <span class="fu">transcripts</span><span class="op">(</span><span class="va">edb</span><span class="op">)</span></span></code></pre></div>
<p>We can group to the gene level with the following
<code>group_id</code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">plyranges</span><span class="op">)</span></span>
<span><span class="va">txps</span> <span class="op">&lt;-</span> <span class="va">txps</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">tx_id</span>, group_id<span class="op">=</span><span class="va">gene_id</span><span class="op">)</span></span></code></pre></div>
<p>The following code chunk imports the allelic data, creating a “wide”
<em>SummarizedExperiment</em> object which will have 2x the number of
columns as the samples listed in <code>coldata</code>. We specify the
strings that distinguish the <code>a1</code> and <code>a2</code> allele
in the transcript FASTA file (these are suffices that follow the
transcript name, separated with an underscore,
e.g. <code>ENST123_ref</code> and <code>ENST123_alt</code>. By
convention, the <code>a2</code> allele describes the reference allele.
In the <code>gse</code> object (gene-level SummarizedExperiment), the
<code>a2</code> allelic counts will appear as the first columns followed
by the <code>a1</code> allelic counts. Here we supply <code>txps</code>
as a <em>GRanges</em> object with a metadata column
<code>group_id</code> that describes how we collapse transcript-level
counts.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/fishpond" class="external-link">fishpond</a></span><span class="op">)</span></span>
<span><span class="va">gse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/fishpond/man/importAllelicCounts.html" class="external-link">importAllelicCounts</a></span><span class="op">(</span></span>
<span>  <span class="va">coldata</span>, a1<span class="op">=</span><span class="st">"alt"</span>, a2<span class="op">=</span><span class="st">"ref"</span>,</span>
<span>  format<span class="op">=</span><span class="st">"wide"</span>, tx2gene<span class="op">=</span><span class="va">txps</span>,</span>
<span><span class="op">)</span></span></code></pre></div>
<p>Finally, we performed minimal filtering on the features:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># filtering out lowly expressed features:</span></span>
<span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">gse</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">6</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">keep</span><span class="op">)</span></span>
<span><span class="va">gse</span> <span class="op">&lt;-</span> <span class="va">gse</span><span class="op">[</span><span class="va">keep</span>,<span class="op">]</span></span></code></pre></div>
<p>Alternatively, we can group transcript-level counts to the
transcription start site (TSS) level using the following code chunk to
define <code>txps</code>. <code><a href="https://rdrr.io/pkg/fishpond/man/makeTx2Tss.html" class="external-link">makeTx2Tss()</a></code> is a convenience
function in <em>fishpond</em>. The <code>maxgap=50</code> argument means
we will group together transcripts that have TSS that fall within 50bp
of each other.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">txps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/fishpond/man/makeTx2Tss.html" class="external-link">makeTx2Tss</a></span><span class="op">(</span><span class="va">edb</span>, maxgap<span class="op">=</span><span class="fl">50</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">tx_id</span>, <span class="va">gene_id</span>, <span class="va">group_id</span>, <span class="va">tss</span><span class="op">)</span></span></code></pre></div>
<p><strong>Q: why would we want to collapse to TSS level?</strong></p>
</div>
<div class="section level2">
<h2 id="mouse-osteoblast-differentiation">Mouse osteoblast differentiation<a class="anchor" aria-label="anchor" href="#mouse-osteoblast-differentiation"></a>
</h2>
<p>We have two pre-packaged datasets as part of this workshop package,
one where allelic counts from an RNA-seq experiment are summarized to
the gene level, and another where they are summarized to the TSS
level.</p>
<p>The datasets are described in the man pages (see References tab).
Briefly, mouse osteoblast cells were differentiated over a time course
from day 2 to day 18. The data has been subset to just the genes on
chr1. A reference for the experiment is:</p>
<blockquote>
<p>Kemp JP, Medina-Gomez C, Estrada K, St Pourcain B et al. Phenotypic
dissection of bone mineral density reveals skeletal site specificity and
facilitates the identification of novel loci in the genetic regulation
of bone mass attainment. PLoS Genet 2014 Jun;10(6):e1004423. PMID:
24945404 <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4063697/" class="external-link uri">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4063697/</a></p>
</blockquote>
<p>…more about osteoblasts and their genetic importance…</p>
<p><strong>Q: why would we be interested in allelic expression during
osteoblast differentiation?</strong></p>
</div>
<div class="section level2">
<h2 id="explore-allelic-counts">Explore allelic counts<a class="anchor" aria-label="anchor" href="#explore-allelic-counts"></a>
</h2>
<p>We load the gene-level dataset created with
<code><a href="https://rdrr.io/pkg/fishpond/man/importAllelicCounts.html" class="external-link">importAllelicCounts()</a></code>:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mikelove.github.io/Bioc2022AllelicExpression">Bioc2022AllelicExpression</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">osteoblast_gene_chr1</span><span class="op">)</span></span>
<span><span class="va">osteoblast_gene_chr1</span></span>
<span><span class="co">#&gt; class: RangedSummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 1211 36 </span></span>
<span><span class="co">#&gt; metadata(3): tximetaInfo countsFromAbundance alleles</span></span>
<span><span class="co">#&gt; assays(33): counts abundance ... infRep29 infRep30</span></span>
<span><span class="co">#&gt; rownames(1211): ENSMUSG00000001138 ENSMUSG00000001143 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000118219 ENSMUSG00000118607</span></span>
<span><span class="co">#&gt; rowData names(9): gene_id gene_name ... symbol entrezid</span></span>
<span><span class="co">#&gt; colnames(36): 129xB6-d02-a2 129xB6-d04-a2 ... CASTxB6-d16-a1</span></span>
<span><span class="co">#&gt;   CASTxB6-d18-a1</span></span>
<span><span class="co">#&gt; colData names(3): allele cross day</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="va">osteoblast_gene_chr1</span></span></code></pre></div>
<p>The sample metadata:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 36 rows and 3 columns</span></span>
<span><span class="co">#&gt;                  allele    cross       day</span></span>
<span><span class="co">#&gt;                &lt;factor&gt; &lt;factor&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; 129xB6-d02-a2        a2   129xB6         2</span></span>
<span><span class="co">#&gt; 129xB6-d04-a2        a2   129xB6         4</span></span>
<span><span class="co">#&gt; 129xB6-d06-a2        a2   129xB6         6</span></span>
<span><span class="co">#&gt; 129xB6-d08-a2        a2   129xB6         8</span></span>
<span><span class="co">#&gt; 129xB6-d10-a2        a2   129xB6        10</span></span>
<span><span class="co">#&gt; ...                 ...      ...       ...</span></span>
<span><span class="co">#&gt; CASTxB6-d10-a1       a1  CASTxB6        10</span></span>
<span><span class="co">#&gt; CASTxB6-d12-a1       a1  CASTxB6        12</span></span>
<span><span class="co">#&gt; CASTxB6-d14-a1       a1  CASTxB6        14</span></span>
<span><span class="co">#&gt; CASTxB6-d16-a1       a1  CASTxB6        16</span></span>
<span><span class="co">#&gt; CASTxB6-d18-a1       a1  CASTxB6        18</span></span></code></pre></div>
<p>For each sample and each allele, we have counts and also bootstrap
replicates. The bootstrap replicates are from Salmon, by specifying
<code>--numBootstraps 30</code> during quantification.</p>
<p><strong>Q: Why do we need bootstrap replicates?</strong></p>
<p>We can use <em>fishpond</em> functions to do some basic exploration
of the allelic counts.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/fishpond" class="external-link">fishpond</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assayNames</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "counts"    "abundance" "length"    "infRep1"   "infRep2"   "infRep3"  </span></span>
<span><span class="co">#&gt;  [7] "infRep4"   "infRep5"   "infRep6"   "infRep7"   "infRep8"   "infRep9"  </span></span>
<span><span class="co">#&gt; [13] "infRep10"  "infRep11"  "infRep12"  "infRep13"  "infRep14"  "infRep15" </span></span>
<span><span class="co">#&gt; [19] "infRep16"  "infRep17"  "infRep18"  "infRep19"  "infRep20"  "infRep21" </span></span>
<span><span class="co">#&gt; [25] "infRep22"  "infRep23"  "infRep24"  "infRep25"  "infRep26"  "infRep27" </span></span>
<span><span class="co">#&gt; [31] "infRep28"  "infRep29"  "infRep30"</span></span>
<span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/fishpond/man/getTrace.html" class="external-link">getTrace</a></span><span class="op">(</span><span class="va">g</span>, idx<span class="op">=</span><span class="fl">1</span>, samp_idx<span class="op">=</span><span class="st">"129xB6-d02-a2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">count</span>, border<span class="op">=</span><span class="st">"white"</span>, col<span class="op">=</span><span class="st">"grey50"</span>, xlab<span class="op">=</span><span class="st">"estimated count, a2"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Bioc2022AllelicExpression_files/figure-html/histBoot-1.png" width="700"></p>
<p>This plot shows the distribution of estimated counts for the first
feature, and for the day 2 sample of the 129xB6 cross, reference allele.
We can also visualize the distributions for all samples using
<code><a href="https://rdrr.io/pkg/fishpond/man/plotInfReps.html" class="external-link">plotInfReps()</a></code>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/fishpond/man/plotInfReps.html" class="external-link">plotInfReps</a></span><span class="op">(</span><span class="va">g</span>, idx<span class="op">=</span><span class="fl">1</span>, x<span class="op">=</span><span class="st">"allele"</span>, legend<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Bioc2022AllelicExpression_files/figure-html/plotInfReps-1.png" width="700"></p>
<p>We can also group the allelic counts ordered by sample, e.g.:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gcast</span> <span class="op">&lt;-</span> <span class="va">g</span><span class="op">[</span>,<span class="va">g</span><span class="op">$</span><span class="va">cross</span> <span class="op">==</span> <span class="st">"CASTxB6"</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/fishpond/man/plotInfReps.html" class="external-link">plotInfReps</a></span><span class="op">(</span><span class="va">gcast</span>, idx<span class="op">=</span><span class="fl">1</span>, x<span class="op">=</span><span class="st">"allele"</span>, cov<span class="op">=</span><span class="st">"day"</span>, legend<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Bioc2022AllelicExpression_files/figure-html/plotInfReps2-1.png" width="700"></p>
<p>In the above plot, the samples are ordered by sample, as if
<code>day</code> x <code>allele</code> were a categorical variable. We
can also plot <code>day</code> on its original continuous scale. To do
so, we use points and lines to represent the bootstrap distribution
instead of boxplots. To draw the line, we need to first compute the
bootstrap variance, which can be computed with the
<code><a href="https://rdrr.io/pkg/fishpond/man/computeInfRV.html" class="external-link">computeInfRV()</a></code> function (InfRV = “inferential relative
variance”, a useful diagnostic for comparing uncertainty across
features):</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gcast</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/fishpond/man/computeInfRV.html" class="external-link">computeInfRV</a></span><span class="op">(</span><span class="va">gcast</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/fishpond/man/plotInfReps.html" class="external-link">plotInfReps</a></span><span class="op">(</span><span class="va">gcast</span>, idx<span class="op">=</span><span class="fl">1</span>, x<span class="op">=</span><span class="st">"day"</span>, cov<span class="op">=</span><span class="st">"allele"</span>,</span>
<span>            shiftX<span class="op">=</span><span class="fl">.2</span>, legend<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Bioc2022AllelicExpression_files/figure-html/plotInfReps3-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="filter-non-informative-features">Filter non-informative features<a class="anchor" aria-label="anchor" href="#filter-non-informative-features"></a>
</h2>
<p>Some features have no sequence difference between the two alleles.
One easy way to detect these is to look at a single bootstrap replicate,
e.g. the first one <code>infRep1</code>. If this bootstrap sample gives
the same count to each allele, across all samples, this implies either
that the sequence was identical across alleles, or there were no reads
covering the part of the transcript sequence that contains a difference
across alleles. We remove these from the dataset, as they are
non-informative for allelic imbalance analysis.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">gcast</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span></span>
<span><span class="va">rep1a1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">gcast</span>, <span class="st">"infRep1"</span><span class="op">)</span><span class="op">[</span>, <span class="va">gcast</span><span class="op">$</span><span class="va">allele</span> <span class="op">==</span> <span class="st">"a1"</span> <span class="op">]</span></span>
<span><span class="va">rep1a2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">gcast</span>, <span class="st">"infRep1"</span><span class="op">)</span><span class="op">[</span>, <span class="va">gcast</span><span class="op">$</span><span class="va">allele</span> <span class="op">==</span> <span class="st">"a2"</span> <span class="op">]</span></span>
<span><span class="va">someInfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">rep1a1</span> <span class="op">-</span> <span class="va">rep1a2</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">1</span><span class="op">)</span> <span class="op">&lt;</span> <span class="va">n</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">someInfo</span><span class="op">)</span></span>
<span><span class="co">#&gt; someInfo</span></span>
<span><span class="co">#&gt; FALSE  TRUE </span></span>
<span><span class="co">#&gt;    50  1161</span></span>
<span><span class="va">gcast</span> <span class="op">&lt;-</span> <span class="va">gcast</span><span class="op">[</span> <span class="va">someInfo</span>, <span class="op">]</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Euphy Wu, Michael Love.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
